mpmp/prediction/classification.py

```diff
@@ -233,32 +234,156 @@ def train_gb_classifier(X_train,
             )
         ]
     )
-
-    cv_pipeline = GridSearchCV(
-        estimator=estimator,
-        param_grid=clf_parameters,
-        n_jobs=-1,
-        cv=n_folds,
-        scoring='average_precision',
-        return_train_score=True,
-    )
+    if n_folds == -1:
+        # for this option we just want to do a grid search for a single
+        # train/test split, this is much more computationally efficient
+        # but could have higher variance
+        from sklearn.model_selection import train_test_split
+        train_ixs, valid_ixs = train_test_split(
+            np.arange(X_train.shape[0]),
+            test_size=cfg.inner_valid_prop,
+            random_state=seed,
+            shuffle=True
+        )
+        cv = zip([train_ixs], [valid_ixs])
+        cv_pipeline = GridSearchCV(
+            estimator=estimator,
+            param_grid=clf_parameters,
+            n_jobs=-1,
+            cv=cv,
+            scoring='average_precision',
+            return_train_score=True,
+        )
+    else:
+        cv_pipeline = GridSearchCV(
+            estimator=estimator,
+            param_grid=clf_parameters,
+            n_jobs=-1,
+            cv=n_folds,
+            scoring='average_precision',
+            return_train_score=True,
+        )
 
     # Fit the model
     cv_pipeline.fit(X=X_train, y=y_train.status)
 
-    # Obtain cross validation results
-    y_cv = cross_val_predict(
-        cv_pipeline.best_estimator_,
-        X=X_train,
-        y=y_train.status,
-        cv=n_folds,
-        method='predict_proba',
-    )[:, 1]
-
     # Get all performance results
     y_predict_train = cv_pipeline.predict_proba(X_train)[:, 1]
     y_predict_test = cv_pipeline.predict_proba(X_test)[:, 1]
 
+    # Obtain cross validation results, if applicable
+    if n_folds == -1:
+        y_cv = y_predict_train
+    else:
+        y_cv = cross_val_predict(
+            cv_pipeline.best_estimator_,
+            X=X_train,
+            y=y_train.status,
+            cv=n_folds,
+            method='decision_function',
+        )
+
+    return cv_pipeline, y_predict_train, y_predict_test, y_cv
+
+
+def train_mlp_classifier(X_train,
+                         X_test,
+                         y_train,
+                         params,
+                         seed,
+                         n_folds=4,
+                         max_iter=1000):
+    """Train multi-layer perceptron classifier."""
+    import torch.optim
+    from torch.utils.data import Dataset
+    from skorch import NeuralNetBinaryClassifier
```

Comment:
Haha, yeah - it's really useful if you're just implementing a basic model like I'm doing here. It was nice to avoid writing all the training/test loop and cross-validation logic myself, especially since I'm already using a lot of sklearn.

For you, though, since you want to do weird stuff and have a lot of options for customization, just going with pure pytorch is probably the safer bet.